src/uct/tcp/tcp_ep.c

```diff
@@ -65,14 +65,184 @@ static inline int uct_tcp_ep_ctx_buf_need_progress(uct_tcp_ep_ctx_t *ctx)
+    if (status != UCS_OK) {
+        goto err;
+    }
+
+    status = uct_tcp_iface_set_sockopt(iface, ep->fd);
+    if (status != UCS_OK) {
+        goto err;
+    }
+
+    uct_tcp_cm_conn_start(ep);
+
+out:
+    return 1;
+
+err:
+    uct_tcp_ep_set_failed(ep);
+    goto out;
+}
+
+static ucs_status_t
+uct_tcp_ep_create_socket_and_connect(uct_tcp_ep_t *ep)
+{
+    uct_tcp_iface_t *iface   = ucs_derived_of(ep->super.super.iface,
+                                              uct_tcp_iface_t);
+    uct_worker_cb_id_t cb_id = UCS_CALLBACKQ_ID_NULL;
+
+
+    if (ep->conn_retries++ > iface->config.max_conn_retries) {
+        ucs_error("tcp_ep %p: reached maximum number of connection retries "
+                  "(%u)", ep, iface->config.max_conn_retries);
+        uct_tcp_ep_set_failed(ep);
+        return UCS_ERR_TIMED_OUT;
+    }
+
+    uct_worker_progress_register_safe(&iface->super.worker->super,
+                                      uct_tcp_ep_create_socket_and_connect_progress,
+                                      ep, UCS_CALLBACKQ_FLAG_ONESHOT, &cb_id);
+
+    return UCS_OK;
+}
+
+int uct_tcp_ep_is_unconnected(uct_tcp_ep_t *ep)
+{
+    return (ep->conn_state == UCT_TCP_EP_CONN_STATE_CLOSED) &&
+           (ep->conn_retries == 0);
+}
+
+int uct_tcp_ep_is_connect_in_progress(uct_tcp_ep_t *ep)
+{
+    return (ep->conn_state == UCT_TCP_EP_CONN_STATE_CONNECTING) &&
+           (ep->conn_retries == 0);
+}
+
+static unsigned uct_tcp_ep_connect_progress(void *arg)
+{
+    uct_tcp_ep_t *ep       = (uct_tcp_ep_t*)arg;
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+    uct_tcp_ep_t *peer_ep;
+    ucs_status_t status;
+
+    ucs_assert(uct_tcp_ep_is_connect_in_progress(ep));
+
+    if (uct_tcp_ep_is_self(ep) ||
+        (peer_ep = uct_tcp_cm_get_ep(iface, &ep->peer_addr, ep->peer_conn_sn,
+                                     UCT_TCP_EP_FLAG_CTX_TYPE_RX)) == NULL) {
+        uct_tcp_ep_create_socket_and_connect(ep);
+        goto out;
+    }
+
+    uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_CONNECTED);
+
+    ucs_assert(!(peer_ep->flags & UCT_TCP_EP_FLAG_CTX_TYPE_TX) &&
+               (peer_ep->fd != -1));
+    uct_tcp_ep_move_ctx_cap(peer_ep, ep, UCT_TCP_EP_FLAG_CTX_TYPE_RX);
+
+    ep->fd          = peer_ep->fd;
+    ep->events      = peer_ep->events;
+    peer_ep->fd     = -1;
+    peer_ep->events = 0;
+
+    ep->conn_retries++;
+
+    uct_tcp_ep_ctx_move(&ep->tx, &peer_ep->tx);
+    uct_tcp_ep_ctx_move(&ep->rx, &peer_ep->rx);
+
+    ucs_queue_splice(&ep->pending_q, &peer_ep->pending_q);
+    ucs_queue_splice(&ep->put_comp_q, &peer_ep->put_comp_q);
+
+    /* The internal EP is not needed anymore */
+    uct_tcp_ep_set_failed(peer_ep);
+    peer_ep = NULL;
+
+    /* Send the connection request to the peer */
+    status = uct_tcp_cm_send_event(ep, UCT_TCP_CM_CONN_REQ, 0);
+    if (status != UCS_OK) {
+        goto out;
```

Comment:
don't we need to return here?

Reply:
yes, no need to add to matching context
but need to set TX cap